grease/src/Grease/Shape.hs

```diff
@@ -397,3 +411,84 @@ replaceShapes names (ArgShapes args) (ParsedShapes replacements) =
                         , foundType = C.Some ty'
                         }
       Nothing -> Right s
+
+-- | Given a parser for @tag@s and 'ShapeExt's, parse 'Shape's from JSON
+parseJsonShape ::
+  -- | Parser for @tag@s
+  (forall t. Aeson.Value -> Aeson.Parser (tag t)) ->
+  -- | Parser for 'ExtShape's
+  (Aeson.Value -> Aeson.Parser (Some (ExtShape ext tag))) ->
+  -- | JSON value to parse
+  Aeson.Value ->
+  Aeson.Parser (Some (Shape ext tag))
+parseJsonShape parseTag parseExt =
+  Aeson.withObject "Shape" $ \v -> do
+    ty <- v .: "type" :: Aeson.Parser Text
+    let tg = v .: "tag"
+    case ty of
+      "bool" -> Some . ShapeBool <$> (parseTag =<< tg)
+      "ext" -> doParseExt v
+      "float" -> parseFloat tg v
+      "struct" -> parseStruct tg v
+      "unit" -> Some . ShapeUnit <$> (parseTag =<< tg)
+      _ -> fail ("Unknown Shape type: " ++ Text.unpack ty)
+ where
+  doParseExt v = do
+    Some s <- parseExt =<< v .: "val"
+    pure (Some (ShapeExt s))
+
+  parseFloat tg v = do
+    Some fi <- parseFloatRepr =<< v .: "info"
+    tag <- parseTag =<< tg
```

Comment:
do we even want to be parsing the tag, right now it could be anything and we will const it, seems a bit odd to me, it's not clear when we would ever want things tagged?

Reply:
yeah i guess that's fair but imo if there's no tag we shouldnt be requiring "tag": 1 so imo when the type is NoTag ideally we wouldnt be fetching the tg at all